README.rst
LKKanban: LeanKitKanban's API for Python
========================================

LKKanban aspires to make it easy to interact with LeanKitKanban via python.
Unlike other python LeanKit libraries, it is permissively licensed (it uses
ISC). Although it aspires to provide a pythonic API, right now it is just a
simple wrapper around the HTTP API, using Kenneth Reitz's excellent
`Requests <http://docs.python-requests.org/en/latest/index.html>`_. In fact,
it currently only gets some data -- it cannot update a board. That should
change soon though. See lkkanban.http_api.HttpApi for which endpoints are
currently wrapped, until they all are (at which point this sentence will
go away).

Using the HTTP API wrapper looks like this.

::

    >>> from lkkanban.http_api import HttpApi
    >>> api = HttpApi('account', 'user@example.com', 'password')
    >>> api.get_boards()
    {u'ReplyText': u'Board(s) successfully retrieved.', u'ReplyCode': 200,
    u'ReplyData': [[{u'BreakoutBoards': None, u'Description': u'', u'Title':
    u'Our Board', u'DrillThroughBoards': [], u'IsBreakoutBoard': False,
    u'ParentId': 0, u'IsArchived': False, u'CreationDate': u'20/01/2012',
    'Id': 10000000}]]}

API method names follow the `Lean Kit docs
<http://support.leankitkanban.com/forums/20153741-api>`_, but with the
CamelCase method names converted to PEP8 compliant names (so
AddCardWithWipOverride becomes add_card_with_wip_override).

All methods return JSON.

Any method that should post takes a dictionary with the request body, with all
names following LeanKit's docs.

Developed by `Matt Bowen <http://www.mattbowen.net>`_.

HISTORY.rst
History
=======

0.1 (unreleased)
++++++++++++++++

- Wrap HTTP API of LeanKitKanban

lkkanban/tests/__init__.py
# Intentionall blank


lkkanban/tests/test_http_api.py
import unittest
from lkkanban.httpapi import HttpApi


class TestHttpApi(unittest.TestCase):
    def setUp(self):
        self.local = HttpApi('foo', 'username', 'passwd', 'fixtures')
        self.remote = HttpApi('foo', 'username', 'passwd')

    def test_get_endpoint_all_boards(self):
        expected_endpoint = 'http://foo.leankitkanban.com/Kanban/Api/Boards'
        actual_endpoint = self.remote._get_endpoint('Boards')
        self.assertEqual(expected_endpoint, actual_endpoint)

    def test_get_endpoint_addcard(self):
        expected_endpoint = \
            "http://foo.leankitkanban.com/Kanban/Api/Board/100/AddCards"
        actual_endpoint = self.remote._get_endpoint('AddCards', 100)
        self.assertEqual(expected_endpoint, actual_endpoint)

    def test_get_endpoint_specific_board(self):
        expected_endpoint = \
            "http://foo.leankitkanban.com/Kanban/Api/Boards/100"
        actual_endpoint = self.remote._get_endpoint('Boards', 100)
        self.assertEqual(expected_endpoint, actual_endpoint)


lkkanban/httpapi.py
import requests


class HttpApi(object):
    """A minimal wrapper around the LeanKitKanban HTTP API.
       Methods take dictionaries for POST data, returns JSON
    """
    def __init__(self, account_name, username, password, _service_base=None):
        self.account_name = account_name
        self.session = requests.session(auth=(username, password))
        self._service_base = _service_base

    def _get_endpoint(self, name, board_id=''):
        """All LeanKit endpoints include the account name, so this will
        take an endpoint name and construct a leankit endpoint URL with
        the account name included"""
        url_template = \
            ("http://%(account)s.leankitkanban.com/Kanban/"
             "Api/%(board_id)s$%(name)s")
        if name == 'Boards':
            url_template = url_template.replace(
                '%(board_id)s$%(name)s', '%(name)s$%(board_id)s')
        elif board_id != '':
            board_id = ''.join(('Board/', str(board_id)))
        if board_id == '':
            url_template = url_template.replace('$', '')
        else:
            url_template = url_template.replace('$', '/')
        return url_template % \
            {'account': self.account_name, 'name': name, 'board_id': board_id}

    def get_boards(self):
        """Wraps LeanKits' GetBoards method
        See http://support.leankitkanban.com/entries/20264797-get-boards
        """
        endpoint = self._get_endpoint('Boards')
        r = self.session.get(endpoint)
        return r.json

    def get_board_identifiers(self, board_id):
        """Wraps LeanKit's GetBoardIdentifiers
        See http://support.leankitkanban.com/entries/20267921-getboardidentifiers
        """
        endpoint = self._get_endpoint('GetBoardIdentifiers', board_id)
        r = self.session.get(endpoint)
        return r.json

    def get_board(self, board_id):
        """Wraps LeanKit's GetBoard
        See http://support.leankitkanban.com/entries/20267956-get-board
        """
        endpoint = self._get_endpoint('Boards', board_id)
        r = self.session.get(endpoint)
        return r.json

    def get_newer_if_exists(self, board_id, board_version):
        """Wraps LeanKit's GetNewerIfExists
        See http://support.leankitkanban.com/entries/20267966-getnewerifexists
        """
        base_endpoint = self._get_endpoint('BoardVersion', board_id)
        endpoint = '/'.join((
            base_endpoint,
            str(board_version),
            'GetNewerIfExists'))
        r = self.session.get(endpoint)
        return r.json

    def get_board_history_since(self, board_id, board_version):
        """Wraps LeanKit's GetBoardHistorySince
        See http://support.leankitkanban.com/entries/20267971-getboardhistorysince
        """
        base_endpoint = self._get_endpoint('BoardVersion', board_id)
        endpoint = '/'.join((
            base_endpoint,
            str(board_version),
            'GetBoardHistorySince'))
        r = self.session.get(endpoint)
        return r.json

    def get_card(self, board_id, card_id):
        """Wraps LeanKit's GetCard
        See http://support.leankitkanban.com/entries/20267991-getcard
        """
        base_endpoint = self._get_endpoint('GetCard', board_id)
        endpoint = '/'.join((base_endpoint, str(card_id)))
        r = self.session.get(endpoint)
        return r.json


lkkanban/http_api.py
import requests


class HttpApi(object):
    """A minimal wrapper around the LeanKitKanban HTTP API.
       Methods take dictionaries for POST data, returns JSON
    """
    def __init__(self, account_name, username, password, _service_base=None):
        self.account_name = account_name
        self.session = requests.session(auth=(username, password))
        self._service_base = _service_base

    def _get_endpoint(self, name, board_id=''):
        """All LeanKit endpoints include the account name, so this will
        take an endpoint name and construct a leankit endpoint URL with
        the account name included"""
        url_template = \
            ("http://%(account)s.leankitkanban.com/Kanban/"
             "Api/%(board_id)s$%(name)s")
        if name == 'Boards':
            url_template = url_template.replace(
                '%(board_id)s$%(name)s', '%(name)s$%(board_id)s')
        elif board_id != '':
            board_id = ''.join(('Board/', str(board_id)))
        if board_id == '':
            url_template = url_template.replace('$', '')
        else:
            url_template = url_template.replace('$', '/')
        return url_template % \
            {'account': self.account_name, 'name': name, 'board_id': board_id}

    def get_boards(self):
        """Wraps LeanKits' GetBoards method
        See http://support.leankitkanban.com/entries/20264797-get-boards
        """
        endpoint = self._get_endpoint('Boards')
        r = self.session.get(endpoint)
        return r.json

    def get_board_identifiers(self, board_id):
        """Wraps LeanKit's GetBoardIdentifiers
        See http://support.leankitkanban.com/entries/20267921-getboardidentifiers
        """
        endpoint = self._get_endpoint('GetBoardIdentifiers', board_id)
        r = self.session.get(endpoint)
        return r.json

    def get_board(self, board_id):
        """Wraps LeanKit's GetBoard
        See http://support.leankitkanban.com/entries/20267956-get-board
        """
        endpoint = self._get_endpoint('Boards', board_id)
        r = self.session.get(endpoint)
        return r.json

    def get_newer_if_exists(self, board_id, board_version):
        """Wraps LeanKit's GetNewerIfExists
        See http://support.leankitkanban.com/entries/20267966-getnewerifexists
        """
        base_endpoint = self._get_endpoint('BoardVersion', board_id)
        endpoint = '/'.join((
            base_endpoint,
            str(board_version),
            'GetNewerIfExists'))
        r = self.session.get(endpoint)
        return r.json

    def get_board_history_since(self, board_id, board_version):
        """Wraps LeanKit's GetBoardHistorySince
        See http://support.leankitkanban.com/entries/20267971-getboardhistorysince
        """
        base_endpoint = self._get_endpoint('BoardVersion', board_id)
        endpoint = '/'.join((
            base_endpoint,
            str(board_version),
            'GetBoardHistorySince'))
        r = self.session.get(endpoint)
        return r.json

    def get_card(self, board_id, card_id):
        """Wraps LeanKit's GetCard
        See http://support.leankitkanban.com/entries/20267991-getcard
        """
        base_endpoint = self._get_endpoint('GetCard', board_id)
        endpoint = '/'.join((base_endpoint, str(card_id)))
        r = self.session.get(endpoint)
        return r.json

    def get_card_by_external_id(self, board_id, extneral_id):
        """Wraps LeanKit's GetCardByExternalId
        See http://support.leankitkanban.com/entries/20268001-getcardbyexternalid
        """
        base_endpoint = self._get_endpoint('GetCardByExternalId', board_id)
        endpoint = '/'.join((base_endpoint, str(extneral_id)))
        r = self.session.get(endpoint)
        return r.json

